### ISSUE TITLE
Question about MTRNet in Related_Networks

### ISSUE THREAD

**Issue by hsl323** [neouyghur/One-stage-Mask-based-Scene-Text-Eraser#3]
I tried to test MTRNet in Related_Networks, but I found something different from MTRNet paper.

MTRNet paper mentioned that MTRNet is based on U-net. but network in Related_Networks dose not have skip connection and seems to have a slightly different structure.

There is a comparison with MTRNet on MTRNet++ paper, and I wonder if MTRNet has been tested with the structure that exists in the Related_Networks folder.

**Comment by neouyghur** [comment]
@hsl323 the original MTRNet is written with tensorflow. It did use U-Net as backbone. The core idea of MTRNet is using mask as condition. You can used different network as backbone. However, selecting U-Net will be a good option. The following code is original U-Net based MTRNet model:
```
# Author: neouyghur, osmanjan.t@gmail.com
# Data: April, 2018

from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import tensorflow as tf
from tensorflow.python.ops import array_ops
from config_gan import MaskGanConfig
config = MaskGanConfig()

slim = tf.contrib.slim

def u_net(image, phase_train=True, name_prefix='u-net', reuse=None):
    batch_norm_params = {
        # Decay for the moving averages.
        'decay': config.G_BATCH_NORM_DECAY,
        # epsilon to prevent 0s in variance.
        'epsilon': config.G_BATCH_NORM_EPSILON,
        # force in-place updates of mean and variance estimates
        'updates_collections': tf.GraphKeys.UPDATE_OPS,
        # Moving averages ends up in the trainable variables collection
        # 'variables_collections': [tf.GraphKeys.MOVING_AVERAGE_VARIABLES],
    }
    gf_dim = config.G_BASE_FILTER_COUNT
    with tf.variable_scope(name_prefix, reuse=reuse):
        with slim.arg_scope([slim.batch_norm], is_training=phase_train):
            with slim.arg_scope([slim.conv2d, slim.conv2d_transpose],
                                weights_initializer=tf.truncated_normal_initializer(stddev=config.G_STDDEV),
                                weights_regularizer=slim.l2_regularizer(config.G_REGULARIZE_LAMDA),
                                normalizer_fn=slim.batch_norm,
                                normalizer_params=batch_norm_params):
                with slim.arg_scope([slim.conv2d], stride=2, padding='SAME', activation_fn=tf.nn.leaky_relu):
                    # slim.conv2d default relu activation
                    # subsampling
                    conv0 = slim.conv2d(image, gf_dim, [4,4], scope='conv0')
                    conv1 = slim.conv2d(conv0, gf_dim*2, [4,4], scope='conv1')
                    conv2 = slim.conv2d(conv1, gf_dim*4, [4,4], scope='conv2')
                    conv3 = slim.conv2d(conv2, gf_dim*8, [4,4], scope='conv3')
                    conv4 = slim.conv2d(conv3, gf_dim*8, [4,4], scope='conv4')
                    conv5 = slim.conv2d(conv4, gf_dim*8, [4,4], scope='conv5')
                    conv6 = slim.conv2d(conv5, gf_dim*8, [4,4], scope='conv6')

                with slim.arg_scope([slim.conv2d_transpose], stride=2, padding='SAME'):
                    with slim.arg_scope([slim.conv2d], stride=1, padding='SAME'):
                        # upsampling
                        conv_t0 = slim.conv2d_transpose(conv6, gf_dim*8, [4, 4], scope='conv_t0')
                        merge0 = tf.concat([conv_t0, conv5], 3)
                        conv7 = slim.stack(merge0, slim.conv2d, [(gf_dim*8, [4,4])], scope='conv7')

                        conv_t1 = slim.conv2d_transpose(conv7, gf_dim*8, [2, 2], scope='conv_t1')
                        merge1 = tf.concat([conv_t1, conv4], 3)
                        conv8 = slim.stack(merge1, slim.conv2d, [(gf_dim*8, [4,4])], scope='conv8')

                        conv_t2 = slim.conv2d_transpose(conv8, gf_dim*8, [2, 2], scope='conv_t2')
                        merge2 = tf.concat([conv_t2, conv3], 3)
                        conv9 = slim.stack(merge2, slim.conv2d, [(gf_dim*8, [4,4])], scope='conv9')

                        conv_t3 = slim.conv2d_transpose(conv9, gf_dim*4 , [2, 2], scope='conv_t3')
                        merge3 = tf.concat([conv_t3, conv2], 3)
                        conv10 = slim.stack(merge3, slim.conv2d, [(gf_dim*4, [4, 4])], scope='conv10')

                        conv_t4 = slim.conv2d_transpose(conv10, gf_dim*2, [2, 2], scope='conv_t4')
                        merge4 = tf.concat([conv_t4, conv1], 3)
                        conv11 = slim.stack(merge4, slim.conv2d, [(gf_dim*2, [4, 4])], scope='conv11')

                        conv_t5 = slim.conv2d_transpose(conv11, gf_dim , [2, 2], scope='conv_t5')
                        merge5 = tf.concat([conv_t5, conv0], 3)
                        conv12 = slim.stack(merge5, slim.conv2d, [(gf_dim, [4, 4])], scope='conv12')

                        # different from keras implementation
                        conv_t6 = slim.conv2d_transpose(conv12, gf_dim, [2, 2], scope='conv_t6')
                        conv13 = slim.stack(conv_t6, slim.conv2d, [(gf_dim, [4, 4])], scope='conv13')

                        # output layer scoreMap
                        conv14 = slim.conv2d(conv13, 3, [4, 4], scope='conv14',
                                            activation_fn=tf.nn.tanh, normalizer_fn=None, stride=1)
                        return conv14

def discriminator(input, phase_train=True, reuse=None, name_prefix='discriminator'):
    batch_norm_params = {
        # Decay for the moving averages.
        'decay': config.D_BATCH_NORM_DECAY,
        # epsilon to prevent 0s in variance.
        'epsilon': config.D_BATCH_NORM_EPSILON,
        # force in-place updates of mean and variance estimates
        'updates_collections': tf.GraphKeys.UPDATE_OPS,
        # # Moving averages ends up in the trainable variables collection
        # 'variables_collections': [tf.GraphKeys.MOVING_AVERAGE_VARIABLES],
    }

    with tf.variable_scope(name_prefix, reuse=reuse):
        with slim.arg_scope([slim.batch_norm], is_training=phase_train):
            with slim.arg_scope([slim.conv2d, slim.conv2d_transpose],
                                weights_initializer=tf.truncated_normal_initializer(stddev=config.D_STDDEV),
                                weights_regularizer=slim.l2_regularizer(config.D_REGULARIZE_LAMDA),
                                normalizer_fn=slim.batch_norm,
                                normalizer_params=batch_norm_params,
                                activation_fn=tf.nn.leaky_relu):
                df_dim = config.D_BASE_FILTER_COUNT
                with slim.arg_scope([slim.conv2d], stride=2, padding='SAME'):
                    # No batch normalization for first layer
                    conv0 = slim.conv2d(input, df_dim, [4,4], scope='conv0', normalizer_fn=None)
                    conv1 = slim.conv2d(conv0, df_dim * 2, [4,4], scope='conv1')
                    conv2 = slim.conv2d(conv1, df_dim * 4, [4,4], scope='conv2')
                    conv3 = slim.conv2d(conv2, df_dim * 8, [4,4], scope='conv3')
                with slim.arg_scope([slim.conv2d_transpose], stride=2, padding='SAME'):
                    conv_t0 = slim.conv2d_transpose(conv3, df_dim * 8, [4, 4], scope='conv_t0')
                    conv_t1 = slim.conv2d_transpose(conv_t0, df_dim * 4, [4, 4], scope='conv_t1')
                    conv_t2 = slim.conv2d_transpose(conv_t1, df_dim * 2, [4, 4], scope='conv_t2')
                    conv_t3 = slim.conv2d_transpose(conv_t2, df_dim , [4, 4], scope='conv_t3')
                    conv4 = slim.conv2d(conv_t3, 1, [4,4], scope='conv4',
                                      activation_fn=None, stride=1, normalizer_fn=None)

            net = tf.squeeze(conv4, [3], name='conv4/squeezed')

            return net
```
